src/payment/strategies/moneris/moneris-payment-strategy.ts

```diff
@@ -89,29 +141,74 @@ export default class MonerisPaymentStrategy implements PaymentStrategy {
         if (nonce !== undefined) {
             return this._store.dispatch(this._paymentActionCreator.submitPayment({
                 methodId: payment.methodId,
-                paymentData: { nonce },
+                paymentData: { nonce, shouldSaveInstrument, shouldSetAsDefaultInstrument },
             }));
         }
 
         return this._store.getState();
     }
 
-    finalize(): Promise<InternalCheckoutSelectors> {
-        return Promise.reject(new OrderFinalizationNotRequiredError());
-    }
+    protected async executeWithVaulted(payment: OrderPaymentRequestBody): Promise<InternalCheckoutSelectors> {
+        if (this._isHostedPaymentFormEnabled(payment.methodId)) {
+            const form = this._hostedForm;
 
-    deinitialize(): Promise<InternalCheckoutSelectors> {
-        if (this._windowEventListener) {
-            window.removeEventListener('message', this._windowEventListener);
-            this._windowEventListener = undefined;
+            if (!form) {
+                throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+            }
+
+            await form.validate();
+            await form.submit(payment);
+
+            return await this._store.dispatch(this._orderActionCreator.loadCurrentOrder());
+        } else {
+            return await this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
         }
+    }
 
-        if (this._iframe && this._iframe.parentNode) {
-            this._iframe.parentNode.removeChild(this._iframe);
-            this._iframe = undefined;
+    private shouldShowTSVHostedForm(methodId: string): boolean {
+        return (this._isHostedPaymentFormEnabled(methodId) && this._isHostedFieldAvailable());
+    }
+
+    private _isHostedPaymentFormEnabled(methodId: string): boolean {
+        const { paymentMethods: { getPaymentMethodOrThrow } } = this._store.getState();
+        const paymentMethod = getPaymentMethodOrThrow(methodId);
+
+        return paymentMethod.config.isHostedFormEnabled === true;
+    }
+
+    private _isHostedFieldAvailable(): boolean {
+        const options = this._getInitializeOptions();
+
+        return !!options.form?.fields;
+    }
+
+    private _getInitializeOptions(): MonerisPaymentInitializeOptions {
+        if (!this._initializeOptions) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
         }
 
-        return Promise.resolve(this._store.getState());
+        return this._initializeOptions;
+    }
+
+    private _mountCardVerificationfields(formOptions: HostedFormOptions): Promise<HostedForm> {
```

Comment:
The new promise construction is redundant, you can move `async` keyword on `_mountCardVerificationfields ` function and it should be good.

Reply:
if the promise isn't marked as async, the line with `await form.attach();`  will be marked as an error, since the arrow function make it be in a different scope that would no longer be async.

Reply:
Have you tried it removing the promise and putting `async` next to `_mountCardVerificationfields `?

Reply:
yes, that's what I tried, but it didn't work

Reply:
Have you tried this 
```
private async _mountCardVerificationfields(formOptions: HostedFormOptions): Promise<HostedForm> {
        try {
            const { config } = this._store.getState();
            const bigpayBaseUrl = config.getStoreConfig()?.paymentSettings.bigpayBaseUrl;

            if (!bigpayBaseUrl) {
                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
            }

            const form = this._hostedFormFactory.create(bigpayBaseUrl, formOptions);

            await form.attach();

            return form;
        } catch (error) {
            return error;
        }
    }
```

Reply:
that does, work, I've added it to the PR

Reply:
Thanks for updating that.